ntp-proto/src/nts_record.rs

```diff
@@ -366,3 +379,418 @@ impl NtsRecordDecoder {
         }
     }
 }
+
+#[derive(Debug)]
+pub enum KeyExchangeError {
+    UnrecognizedCriticalRecord,
+    BadRequest,
+    InternalServerError,
+    UnknownErrorCode(u16),
+    NoValidProtocol,
+    NoValidAlgorithm,
+    NoCookies,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct KeyExchange {
+    pub remote: String,
+    pub port: u16,
+    pub cookies: Vec<Vec<u8>>,
+}
+
+pub type KeyExchangeStep = ControlFlow<Result<KeyExchange, KeyExchangeError>, KeyExchange>;
+
+pub enum NextStep {
+    NeedMoreInput(KeyExchange),
+    Done(KeyExchange),
+    ReadError(std::io::Error),
+    KeyExchangeError(KeyExchangeError),
```

Comment:
I'd suggest merging these two error types. Externally, we mostly care about that an error happens, not which type precisely.